src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedStream.cs

```diff
@@ -0,0 +1,1098 @@
+
+            int alreadySatisfied = bytesFromBuffer;
+            if (bytesFromBuffer > 0)
+            {
+                count -= bytesFromBuffer;
+                offset += bytesFromBuffer;
+            }
+
+            // So the READ buffer is empty.
+            Contract.Assert(_readLen == _readPos);
+            _readPos = _readLen = 0;
+
+            // If there was anything in the WRITE buffer, clear it.
+            if (_writePos > 0)
+                FlushWrite();
+
+            // If the requested read is larger than buffer size, avoid the buffer and still use a single read:
+            if (count >= _bufferSize)
+            {
+                return _stream.Read(array, offset, count) + alreadySatisfied;
+            }
+
+            // Ok. We can fill the buffer:
+            EnsureBufferAllocated();
+            _readLen = _stream.Read(_buffer, 0, _bufferSize);
+
+            bytesFromBuffer = ReadFromBuffer(array, offset, count);
+
+            // We may have read less than the number of bytes the user asked for, but that is part of the Stream contract.
+            // Reading again for more data may cause us to block if we're using a device with no clear end of stream,
+            // such as a serial port or pipe.  If we blocked here & this code was used with redirected pipes for a process's
+            // standard output, this can lead to deadlocks involving two processes. Additionally, translating one read on the
+            // BufferedStream to more than one read on the underlying Stream may defeat the whole purpose of buffering of the
+            // underlying reads are significantly more expensive.
+
+            return bytesFromBuffer + alreadySatisfied;
+        }
+
+        private Task<int> LastSyncCompletedReadTask(int val)
+        {
+            Task<int> t = _lastSyncCompletedReadTask;
+            Contract.Assert(t == null || t.Status == TaskStatus.RanToCompletion);
+
+            if (t != null && t.Result == val)
+                return t;
+
+            t = Task.FromResult<int>(val);
+            _lastSyncCompletedReadTask = t;
+            return t;
+        }
+
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+            if (buffer.Length - offset < count)
+                throw new ArgumentException("offset/count");
+            Contract.EndContractBlock();
+
+            // Fast path check for cancellation already requested
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<int>(cancellationToken);
+
+            EnsureNotClosed();
+            EnsureCanRead();
```

Comment:
A personal nit about semantics - I think that the EnsureNotClosed and EnsureCanRead should go above the fast path for cancellation requested. One shouldn't be able to try to do something invalid and have the method call succeed since it got cancelled. Not a blocker, just personal preference on my part.